File: src/components/tutorial/data/meta-prompt-optimizer.ts
/**
 * Meta-Prompt Optimizer Tutorial (Informational)
 *
 * Explains the 4th optimizer agent that uses an LLM to iteratively rewrite
 * prompts based on scored evaluation results, producing concrete prompt
 * variants for canary A/B testing. Covers the extraction pipeline, the
 * Generate-Critique-Refine loop, canary integration, the feedback loop
 * with similarity guard and circuit breaker, and the frontend dashboard.
 */

import type { Tutorial } from "../../../types/tutorial";

export const metaPromptOptimizerTutorial: Tutorial = {
  id: "meta-prompt-optimizer",
  title: "Meta-Prompt Optimizer",
  description:
    "Understand the LLM-based prompt rewriting system that analyzes prompt performance, generates improved variants, and validates them through canary A/B testing with a multi-layered feedback loop.",
  duration: "15 minutes",
  difficulty: "advanced",
  mode: "contextual",
  focusPage: "meta-optimizer",
  category: "Architecture",
  tags: [
    "meta-optimizer",
    "prompt-engineering",
    "canary-testing",
    "a-b-testing",
    "feedback-loop",
    "featured",
  ],
  prerequisites: ["getting-started"],
  learningObjectives: [
    "Understand why prompt optimization matters and what problem it solves",
    "Know how prompt samples are extracted and grouped by performance",
    "Understand the LLM-based Generate-Critique-Refine rewriting loop",
    "See how rewrites are validated through robustness checks and similarity guards",
    "Understand the canary A/B testing pipeline that validates prompt changes",
    "Learn how the feedback loop prevents repeating failed approaches",
    "Know the circuit breaker and adaptive cooldown system",
    "Use the Prompt Optimization tab to monitor and trigger optimizations",
  ],
  steps: [
    {
      id: "intro",
      title: "The Prompt Stagnation Problem",
      content: `The meta-optimizer system has three existing optimizers that tune **parameters, rules, and agent config**. But none of them touch the actual **prompt text** sent to AI agents.

**The problem:** When a prompt systematically leads to poor outcomes (e.g., the verifier agent produces unstructured responses because the prompt doesn't specify JSON format), no optimizer adjusts it. Recommendations sit in PENDING status indefinitely because they're abstract parameter changes, not testable artifacts.

**The solution:** The **Meta-Prompt Optimizer** (Agent 4) uses an LLM to analyze prompt performance data, critique underperforming prompts, rewrite them, and validate the rewrites through canary A/B testing. Every change produces a concrete, testable prompt variant.`,
      estimatedDuration: 1,
      tips: [
        "This is the 4th optimizer alongside Pipeline Prompt, Architecture, and Generation Template optimizers",
        "It never auto-applies changes — every rewrite goes through canary testing first",
      ],
    },
    {
      id: "extraction",
      title: "Step 1: Prompt Sample Extraction",
      content: `The optimizer starts by collecting performance data from recent workflow runs.

**Data Sources:**
- \`phase_token_usage\` — records which phase (generation, verification, agentic) was executed per run
- \`learning_outcomes\` — stores the \`composite_agentic_score\` (0.0–1.0) for each run
- \`prompt_registry\` — stores active prompt variants per agent type

**Grouping:** Samples are grouped by **(phase, agent_type)** to produce per-group metrics:

| Metric | What it measures |
|--------|-----------------|
| **mean_score** | Average composite score (0.0–1.0) |
| **failure_rate** | Percentage of runs with status "failed" |
| **mean_iterations** | Average iterations to completion |
| **mean_cost** | Average cost in USD per run |

The worst-performing group (highest failure rate, >= 10 samples, > 30% failure rate) becomes the optimization target.`,
      estimatedDuration: 1,
      details: `**Default Prompt Resolution:**

When no variant is active in the registry, the system falls back to hardcoded default prompts for each agent type:

- **locator** — JSON output format for file mapping with criteria, confidence, target files
- **verifier** — Structured assessment with status (pass/partial/fail/unreachable), confidence, observations
- **implementer** — Criteria-driven code change instructions with locator context
- **agentic_fixer** — Targeted fix instructions based on verification failure output

This ensures the optimizer can always see and critique the actual prompt being used, even before any variants have been created.`,
    },
    {
      id: "rewriter",
      title: "Step 2: LLM-Based Prompt Rewriting",
      content: `The core of the optimizer is a **Generate-Critique-Refine** loop run by an LLM. The optimizer is itself a LoopConfig workflow (same as the other 3 optimizers) that runs asynchronously.

**Setup Steps** load context via HTTP:
1. Optimizer history and baselines
2. Prompt group metrics (from extraction)
3. Active prompt variants
4. **Evolution history** (previous rewrite attempts and their verdicts)
5. Recent workflow outcomes (L1)
6. Agent trace summaries (L0)

**Thinking Styles** rotate across runs to prevent convergence on a single approach:

| Style | Focus |
|-------|-------|
| **Evidence-driven** | Ground changes in specific failure examples |
| **Structural clarity** | Restructure for clear step-by-step flow |
| **Constraint-tightening** | Add explicit guardrails and output format specs |
| **Few-shot exemplar** | Add concrete input/output examples |`,
      estimatedDuration: 2,
      tips: [
        "The optimizer produces [META_PROMPT_REWRITE] markers in its output, just like the other optimizers produce their recommendation markers",
        "Maximum 1 rewrite per run — it focuses on the single worst prompt",
      ],
    },
    {
      id: "validation",
      title: "Step 3: Multi-Layer Validation",
      content: `Before a rewritten prompt enters canary testing, it passes through **three validation gates**:

**Gate 1: Robustness Check** (\`validate_prompt_robustness\`)
Checks for: role definition, output format spec, boundary constraints, minimum length. Prompts with > 2 warnings are rejected.

**Gate 2: Semantic Similarity Guard** (\`trigram_similarity\`)
Computes text similarity against all previously **rejected** prompts for this agent type. If similarity > 90%, the rewrite is blocked — it's too close to something that already failed.

**Gate 3: Content Deduplication** (\`compute_content_hash\`)
SHA256 hash comparison prevents creating duplicate recommendations. If an identical recommendation already exists in pending/canary/applied status, it's skipped.

Only prompts that pass all three gates proceed to variant creation and canary testing.`,
      estimatedDuration: 1,
      details: `**Why the similarity guard matters:**

Without it, the optimizer can oscillate between similar failed approaches:
1. Run 1: "Add JSON output format" → rejected (score dropped)
2. Run 2: "Add structured JSON output" → nearly identical, would also fail
3. Run 3: "Require JSON response format" → again nearly identical

The trigram similarity gate (Dice coefficient on 3-character substrings) catches this and forces the optimizer to try a fundamentally different approach, like adding few-shot examples instead.`,
    },
    {
      id: "canary",
      title: "Step 4: Canary A/B Testing",
      content: `Every validated rewrite creates three linked records:

1. **Prompt Variant** — the new prompt text, stored in \`prompt_registry\` (initially inactive)
2. **Recommendation** — links the variant to the optimizer run with evidence
3. **Canary Rollout** — starts A/B testing at 20% traffic

**How canary works:**
- 20% of future runs for this agent use the new prompt variant
- 80% continue using the current prompt (baseline)
- After **10+ canary runs**, statistical evaluation begins

**Statistical verdict** (shared engine with all optimizers):
- **Promote** — canary success rate significantly better (z-test, p < 0.05)
- **Rollback** — canary success rate significantly worse
- **Continue** — insufficient evidence, extend evaluation

For borderline results with 30+ runs, the system automatically **increases canary traffic** (+10%, up to 50%) to gather data faster.`,
      estimatedDuration: 2,
      tips: [
        "Canary evaluation uses Cohen's h effect size and 95% confidence intervals",
        "Stale canaries (active > 30 days without verdict) are automatically rolled back",
      ],
    },
    {
      id: "feedback-loop",
      title: "Step 5: The Feedback Loop",
      content: `The \`prompt_evolution\` table tracks every rewrite attempt and its outcome, creating a **legible history** of prompt optimization:

| Column | Purpose |
|--------|---------|
| \`parent_variant_id\` | Links to the previous version (evolution chain) |
| \`canary_verdict\` | adopt, reject, or NULL (pending) |
| \`score_before\` / \`score_after\` | Quantifies the impact |
| \`baseline_prompt_hash\` | Detects when the baseline prompt changed in code |
| \`consecutive_rejections\` | Powers the circuit breaker |

**After canary completes:**
- **Adopt** → activate the new variant, mark recommendation APPLIED
- **Reject** → mark INEFFECTIVE, record in evolution history for the next round
- **Inconclusive** → extend canary traffic percentage

The next optimizer run **reads the evolution history** and sees rejected approaches. The LLM prompt explicitly instructs: "Previous attempt tried X, which was rejected. Take a different approach."`,
      estimatedDuration: 2,
      details: `**Baseline Drift Detection:**

The \`baseline_prompt_hash\` column stores a SHA256 hash of the baseline prompt at rewrite time. If someone edits the hardcoded prompt in code while a canary is running, the system detects the drift and warns that canary results may be invalid — you'd be comparing against a prompt that no longer exists.`,
    },
    {
      id: "circuit-breaker",
      title: "The Circuit Breaker",
      content: `What happens when multiple rewrites in a row fail? The system has a **diminishing returns circuit breaker** that prevents wasting tokens on a prompt that may already be near-optimal.

**Adaptive Cooldown** (exponential backoff per agent_type):

| Consecutive Rejections | Cooldown Period |
|----------------------|-----------------|
| 0 | 24 hours (standard) |
| 1 | 3 days |
| 2 | 1 week |
| 3 | 2 weeks |
| 4+ | 4 weeks |

**LLM Awareness:** The circuit breaker status is injected into the optimizer's prompt. After 3+ rejections, the LLM is explicitly told:
- The obvious fixes have all failed
- Consider structural alternatives (wrong agent responsibilities, missing context)
- Or accept that the prompt may already be near-optimal and output ZERO rewrites

This prevents the optimizer from endlessly generating variations of the same failed approach.`,
      estimatedDuration: 1,
      tips: [
        "The cooldown is per agent_type — if 'verifier' is in cooldown, 'implementer' can still be optimized",
        "Manual triggers from the UI still check the active-canary guard but bypass the cooldown",
      ],
    },
    {
      id: "trigger-guards",
      title: "When Does the Optimizer Run?",
      content: `The meta-prompt optimizer runs automatically as part of the periodic trigger cycle (same as the other 3 optimizers). But it has **additional guards**:

**Standard guards** (shared with all optimizers):
1. No optimizer of any type already running
2. Meta-optimizer enabled in settings
3. Threshold met (N completed runs since last optimizer)
4. Source run is not a meta-optimizer/fixer/reflection

**Meta-prompt specific guards:**
5. A prompt group exists with **failure rate > 30%** and **>= 10 samples**
6. **No active canary** for the target agent type
7. **Adaptive cooldown** not exceeded (24h to 4 weeks based on rejection history)
8. **Baseline drift warning** (non-blocking) when hardcoded prompts changed

All guards must pass for the optimizer to launch. Manual triggers from the UI check guards 5-6 but skip the cooldown.`,
      estimatedDuration: 1,
    },
    {
      id: "architecture",
      title: "Architecture Overview",
      content: `The meta-prompt optimizer is built from **6 Rust modules** and **1 React component**:

**Rust Backend:**
- \`prompt_extractor.rs\` — Sample extraction, grouping, metrics, default prompt resolver
- \`meta_prompt_optimizer.rs\` — LoopConfig, setup steps, agentic prompt template
- \`prompt_evolution.rs\` — Evolution CRUD, cooldown, similarity, circuit breaker
- \`parser.rs\` — [META_PROMPT_REWRITE] marker parsing + validation pipeline
- \`trigger.rs\` — Meta-prompt specific guards integrated into optimizer dispatch
- \`canary.rs\` — Verdict → evolution feedback wiring

**Frontend:**
- \`PromptOptimizationTab.tsx\` — Dashboard with groups, canaries, evolution history, diff view

**Database:**
- \`prompt_evolution\` table (SQLite + PG dual-write)
- \`prompt_registry\` table (existing, stores variants)
- \`canary_rollouts\` table (existing, manages A/B tests)

**HTTP API:**
4 endpoints under \`/meta-optimizer/prompt-optimization/\` for status, metrics, evidence, and evolution history.`,
      estimatedDuration: 1,
      details: `**Data Flow:**

\`\`\`
phase_token_usage + learning_outcomes
        ↓ (extract_prompt_samples)
  PromptGroupMetrics (grouped by phase/agent_type)
        ↓ (select worst group > 30% failure rate)
  LLM Generate-Critique-Refine loop
        ↓ (parse [META_PROMPT_REWRITE] markers)
  Robustness → Similarity → Dedup validation
        ↓ (create_variant + create_recommendation)
  PromptVariant + CanaryRollout (20% traffic)
        ↓ (auto_evaluate_canaries after 10+ runs)
  Promote / Rollback / Extend
        ↓ (update_evolution_for_recommendation)
  prompt_evolution history → fed back into next run
\`\`\``,
    },
    {
      id: "dashboard",
      title: "Using the Prompt Optimization Tab",
      content: `The **Prompt Optimization** tab on the Meta-Optimizer page gives you visibility into the system:

**Prompt Groups by Failure Rate** — shows all prompt groups ranked by failure rate (worst first). Each group shows:
- Phase/agent type
- Failure rate (red if > 30%)
- Mean score
- Sample count
- Expandable prompt text (from active variant or hardcoded default)

**Active Canaries** — yellow banner showing in-progress A/B tests with baseline score

**Evolution History** — timeline of all rewrite attempts with:
- Score before → score after (green = improved, red = regressed)
- Verdict badge (pending, adopt, reject)
- Consecutive rejection count
- Expandable critique and changes summary
- **View Diff** button — shows line-level diff between old and new prompt

**Trigger Optimization** button — manually launch the optimizer (checks active-canary guard).`,
      estimatedDuration: 1,
      targetElement: {
        selector: '[data-tutorial-id="prompt-optimization-tab"]',
        highlightType: "border",
        position: "bottom",
      },
    },
    {
      id: "mental-model",
      title: "Mental Model Summary",
      content: `Think of the meta-prompt optimizer as an **automated prompt engineer** with guardrails:

1. **Observe** — extract prompt performance data and find the worst-performing prompt
2. **Hypothesize** — use an LLM to critique the prompt and generate an improved version
3. **Validate** — check robustness, similarity to past failures, and deduplication
4. **Experiment** — run an A/B test (canary) to compare new vs. old
5. **Learn** — record the outcome and feed it back into the next iteration
6. **Backoff** — if improvements keep failing, exponentially reduce attempt frequency

This is a **closed-loop optimization system** — not a one-shot rewriter. Each iteration builds on the knowledge from previous attempts, and the system self-regulates to avoid wasting resources on near-optimal prompts.`,
      estimatedDuration: 1,
      tips: [
        "The system is designed to eventually converge: either it finds a better prompt, or the circuit breaker stops it",
        "All data is in SQLite (source of truth) with PG dual-write for analytics",
      ],
      resources: [
        {
          title: "Session Prompt: 04-meta-prompt-optimizer.md",
          url: "qontinui-dev-notes/session-prompts/04-meta-prompt-optimizer.md",
        },
      ],
    },
  ],
};

File: src-tauri/src/meta_optimizer/meta_prompt_optimizer.rs
//! Meta-Prompt Optimizer (Agent 4)
//!
//! Uses an LLM to iteratively rewrite prompts based on scored evaluation results,
//! producing concrete prompt variants for canary testing. Directly addresses the
//! pending recommendations stagnation issue by generating testable artifacts.
//!
//! Unlike the existing three optimizers which adjust parameters/rules/agent config,
//! this optimizer modifies the actual prompt text sent to the AI.

use std::collections::HashMap;

use crate::step_executor::ExecutionStepConfig;
use crate::unified_workflow_executor::LoopConfig;

/// Thinking style variations for mutation diversity.
/// Rotated across runs to explore different prompt rewriting angles.
const META_PROMPT_THINKING_STYLES: &[&str] = &[
    // Style 0: Evidence-driven
    "**Evidence-driven approach:** Ground every change in specific failure examples. \
     Quote exact error patterns and output fragments from failed runs. The rewrite must \
     directly address each documented failure mode. Prefer surgical fixes over broad rewrites.",
    // Style 1: Structural clarity
    "**Structural clarity approach:** Focus on restructuring the prompt for clarity. \
     Separate concerns into distinct sections. Add explicit step-by-step instructions. \
     Ensure the agent knows exactly what order to process information in.",
    // Style 2: Constraint-tightening
    "**Constraint-tightening approach:** Add explicit constraints and guardrails. \
     Define what the agent must NOT do. Add output format specifications. Add boundary \
     checks that prevent the common failure patterns observed in the data.",
    // Style 3: Few-shot exemplar
    "**Few-shot exemplar approach:** Add concrete input/output examples to the prompt. \
     Include at least one example of correct behavior and one example of incorrect behavior \
     (anti-pattern) that matches the observed failure modes. Examples teach better than rules.",
];

/// Build the LoopConfig for the meta-prompt optimizer.
pub fn build_config(execution_id: &str, workflow_name: &str, style_index: u32) -> LoopConfig {
    let base_url = crate::mcp::types::get_self_base_url_from_env();
    let style =
        META_PROMPT_THINKING_STYLES[(style_index as usize) % META_PROMPT_THINKING_STYLES.len()];
    LoopConfig {
        max_iterations: 5,
        base_prompt: build_agentic_prompt(&base_url, style),
        workflow_name: workflow_name.to_string(),
        workflow_id: format!("meta-opt-metaprompt-{}", execution_id),
        execution_id: execution_id.to_string(),
        targeted_error_ids: Vec::new(),
        starting_iteration: 0,
        run_agentic_first: true,
        artifact_dir: None,
        is_dev_mode: false, // CRITICAL: Prevents cascade
        enable_sweep: false,
        max_sweep_iterations: 0,
        stages: Vec::new(),
        stop_on_failure: false,
        constraint_overrides: HashMap::new(),
        reflection_mode: false,
        provider_override: None,
        model_override: None,
        model_overrides: HashMap::new(),
        stage_index: None,
        max_sessions: Some(3),
        auto_run_generated: false,
        approval_gate: false,
        max_context_tokens: 200_000,
        enforce_token_budget: false,
        cross_workflow_learning: false,
        verification_history: HashMap::new(),
        routing_context: Default::default(),
        project_path: crate::mcp::shared::current_project_path(),
        acceptance_criteria: None,
        multi_agent_mode: false,
        strict_cwd: false,
        tool_tags: Vec::new(),
        use_worktree: false,
        worktree_path: None,
        worktree_branch: None,
        workflow_architecture: None,
        agentic_verification_config: None,
        multi_agent_pipeline_config: None,
        rollback_policy: crate::unified_workflow_executor::RollbackPolicy::None,
        escalation_policy: crate::unified_workflow_executor::blame::EscalationPolicy::default(),
        iteration_diffs: Vec::new(),
        active_canary: None,
        is_canary_run: false,
        phase_timeout_ms: None,
    }
}

/// Build setup steps that load extraction data via curl.
pub fn build_setup_steps() -> Vec<ExecutionStepConfig> {
    let base_url = crate::mcp::types::get_self_base_url_from_env();

    vec![
        // Step 0: Load optimizer context (history, baselines, failure patterns)
        build_api_step(
            "Load optimizer context",
            "GET",
            &format!(
                "{}/meta-optimizer/optimizer-context?type=meta_prompt",
                base_url
            ),
            None,
            Some("optimizer_context"),
        ),
        // Step 1: Load prompt group metrics (extraction results)
        build_api_step(
            "Load prompt group metrics",
            "GET",
            &format!(
                "{}/meta-optimizer/prompt-optimization/group-metrics",
                base_url
            ),
            None,
            Some("prompt_group_metrics"),
        ),
        // Step 2: Load current active prompt variants
        build_api_step(
            "Load active prompt variants",
            "GET",
            &format!("{}/meta-optimizer/prompt-variants", base_url),
            None,
            Some("prompt_variants"),
        ),
        // Step 3: Load prompt evolution history (previous rewrite attempts)
        build_api_step(
            "Load prompt evolution history",
            "GET",
            &format!(
                "{}/meta-optimizer/prompt-evolution?limit=20",
                base_url
            ),
            None,
            Some("prompt_evolution_history"),
        ),
        // Step 4: Load recent workflow outcomes (L1)
        build_api_step(
            "Load recent outcomes for batch scoring (L1)",
            "GET",
            &format!("{}/learning/outcomes?limit=20&tier=l1", base_url),
            None,
            Some("recent_outcomes"),
        ),
        // Step 5: Load agent trace summaries (L0)
        build_api_step(
            "Load pipeline agent trace summaries (L0)",
            "GET",
            &format!(
                "{}/meta-optimizer/agent-trace-aggregates?limit=50&tier=l0",
                base_url
            ),
            None,
            Some("agent_trace_aggregates"),
        ),
    ]
}

/// Build verification steps.
pub fn build_verification_steps() -> Vec<ExecutionStepConfig> {
    vec![{
        let mut step = ExecutionStepConfig {
            step_type: "prompt".to_string(),
            name: Some("Verify meta-prompt optimization produced output".to_string()),
            prompt_content: Some(
                r#"Verify that the meta-prompt optimizer analysis is complete.
DO NOT use any UI Bridge SDK tools — they are not available in optimizer mode.

Check:
1. The AI analyzed prompt group metrics and identified the worst-performing prompt group
2. At least one [META_PROMPT_REWRITE] marker was produced (or an explicit explanation of why no rewrite is needed)
3. Each rewrite includes target_phase, target_agent, critique, and rewritten_prompt
4. The rewrite addresses specific failure patterns cited from the evidence data

If the analysis concluded that all prompts are performing adequately and no rewrites are needed, that is acceptable."#
                    .to_string(),
            ),
            ..Default::default()
        };
        step.phase = Some("verification".to_string());
        step
    }]
}

fn build_agentic_prompt(base_url: &str, thinking_style: &str) -> String {
    let prompt = r#"You are the Meta-Prompt Optimizer, part of the meta-optimizer system.

Your job is to analyze actual prompt performance data and rewrite underperforming prompts
to improve their success rate. Unlike other optimizers that adjust parameters or rules,
you directly rewrite the prompt text that is sent to AI agents.

## Data Available

The setup phase loaded the following data:
- `{{optimizer_context}}` — Your performance history: previous recommendations, outcomes, baselines
- `{{prompt_group_metrics}}` — Per-(phase, agent_type) metrics: mean_score, failure_rate, mean_iterations, mean_cost, latest_prompt_text
- `{{prompt_variants}}` — Current active prompt variants in the registry
- `{{prompt_evolution_history}}` — Previous rewrite attempts and their canary verdicts (learn from past failures!)
- `{{recent_outcomes}}` — Last 20 workflow outcomes for context
- `{{agent_trace_aggregates}}` — Per-agent success rates

## Drill-Down Endpoints

If you need more detail on specific failure cases:

**Failed runs for a phase:**
```bash
curl -s '{{base_url}}/meta-optimizer/prompt-optimization/evidence?phase=<PHASE>&max_failures=5&max_successes=2'
```
Returns failure and success samples with task_run_id, outcome_score, outcome_status, iteration, cost_usd.

**Full learning outcome for a specific run:**
```bash
curl -s '{{base_url}}/learning/outcomes/<TASK_RUN_ID>'
```

## THINKING STYLE

{{thinking_style}}

## Prerequisites — Check Before Proceeding

1. **Prompt group metrics exist.** If {{prompt_group_metrics}} is empty, there is no data to optimize.
2. **A group has failure_rate > 0.30.** Only optimize prompts that are clearly underperforming.
3. **No active canary for the target.** Check {{prompt_evolution_history}} — if there's an active canary for the same (phase, agent_type), do NOT create another rewrite.
4. **Learn from history.** Check {{prompt_evolution_history}} for previous rewrite attempts that were rejected. Do NOT repeat the same approach — try a different angle.

If prerequisites are NOT met, explain why and output ZERO [META_PROMPT_REWRITE] markers.

## Learning From Past Failures

**CRITICAL:** The system will programmatically BLOCK any rewrite that is >90% similar to a
previously rejected prompt. You MUST produce a substantially different approach.

Check `{{prompt_evolution_history}}` for entries with `canary_verdict: "reject"`.
For each rejected attempt, the `consecutive_rejections` counter tells you how many
times in a row rewrites have been rejected for that agent.

### REJECTED APPROACHES — DO NOT REPEAT

If `{{prompt_evolution_history}}` contains rejected entries for your target agent,
they will appear below. Study each one carefully:

**For each rejected entry, note:**
- `changes_summary`: What approach was tried
- `score_before` → `score_after`: How the score changed (decrease = regression)
- `critique`: The original diagnosis that led to this failed rewrite

**You MUST:**
1. Explicitly name each prior rejected approach in your output
2. Explain why your new approach is fundamentally different
3. If `consecutive_rejections` >= 3, the obvious fixes have all failed.
   Consider: (a) the prompt may already be near-optimal — output ZERO markers,
   or (b) the problem is structural (wrong agent responsibilities, missing context)
   rather than instructional (prompt wording).
4. Format in your output: "Previous attempts: [list]. This rewrite differs by: [explanation]."

### CIRCUIT BREAKER

If `consecutive_rejections` for an agent reaches 4+, the system has exponentially
increased cooldowns. The prompt is likely near-optimal or has issues that prompt
rewriting cannot fix. Seriously consider outputting ZERO [META_PROMPT_REWRITE] markers
and instead documenting what structural changes might help.

## Your Task

### Step 1: Select Target

From {{prompt_group_metrics}}, identify the prompt group with:
- Highest failure_rate (must be > 0.30)
- Sufficient sample_count (>= 10)
- No active canary already testing a rewrite

### Step 2: Gather Evidence

Use the drill-down endpoint to load failure and success samples:
```bash
curl -s '{{base_url}}/meta-optimizer/prompt-optimization/evidence?phase=<PHASE>&max_failures=5&max_successes=2'
```

Analyze what the failures have in common and what the successes do differently.

### Step 3: Critique the Current Prompt

Identify specific weaknesses in the current prompt (from latest_prompt_text in the group metrics or from the active variant):
- Is the role definition unclear?
- Are constraints missing or too vague?
- Does the prompt lack examples?
- Are there ambiguous instructions?
- Is the output format underspecified?

### Step 4: Rewrite the Prompt

Produce a rewritten prompt that:
- Addresses each identified weakness
- Preserves what works (from success samples)
- Follows the thinking style directive above
- Passes basic robustness checks (has role definition, output format, boundary constraints)

### Step 5: Output

For each rewrite, output a [META_PROMPT_REWRITE] marker:

```
[META_PROMPT_REWRITE]
target_phase: <generation|verification|agentic>
target_agent: <spec_analyst|locator|implementer|verifier|verification|agentic_fixer>
critique: |
  <multi-line critique of the current prompt>
weaknesses: |
  - <weakness 1>
  - <weakness 2>
strengths_to_preserve: |
  - <strength 1>
  - <strength 2>
changes_summary: |
  <brief summary of what changed and why>
expected_improvement: |
  <what metric improvements you expect>
confidence: <0.0-1.0>
rewritten_prompt: |
  <the full rewritten prompt text>
[/META_PROMPT_REWRITE]
```

## Quality Gates

- Maximum 1 rewrite per run (focus on the single worst prompt)
- Minimum confidence: 0.45
- Must cite specific evidence from failure/success samples
- Must explain why the rewrite will improve the specific failure pattern
- Must NOT repeat a previously rejected rewrite approach (check evolution history)
- Rewritten prompt must be a complete replacement, not a diff"#;

    prompt
        .replace("{{base_url}}", base_url)
        .replace("{{thinking_style}}", thinking_style)
}

fn build_api_step(
    name: &str,
    method: &str,
    url: &str,
    body: Option<&str>,
    output_variable: Option<&str>,
) -> ExecutionStepConfig {
    let curl_cmd = if let Some(body_str) = body {
        format!(
            "curl -s -X {} -H \"Content-Type: application/json\" -d '{}' '{}' || echo '{{}}'",
            method, body_str, url
        )
    } else {
        format!("curl -s -X {} '{}' || echo '{{}}'", method, url)
    };

    let extract = output_variable.map(|var| {
        let mut map = HashMap::new();
        map.insert(var.to_string(), "$".to_string());
        map
    });

    let mut step = ExecutionStepConfig {
        step_type: "command".to_string(),
        command_mode: Some("shell".to_string()),
        name: Some(name.to_string()),
        shell_command: Some(curl_cmd),
        extract,
        ..Default::default()
    };
    step.phase = Some("setup".to_string());
    step.run_on_subsequent_iterations = Some(false);
    step
}
